### PR TITLE
feat(fuzz-tests): improve algorithm for minimization

### PR DESCRIPTION
**Pull request by connorskees**
This should add parity to the old implementation in terms of algorithm. It works by applying each transform to each operation until no further changes can be made. This gets us more or less 99% of the way there. The remaining 1% exists in cases where there are two dependent ops -- e.g. if the test case requires that a delete op deletes the exact length of a string inserted immediately prior.

The performance overhead of running these additional transforms is minimal. The most expensive portion of minimization is the initial deletion of unused ops. This is because the execution of a replay involving 100 ops is on the order of 50-100ms. As deletion occurs, the input shrinks and becomes significantly faster to execute.

### CODE REVIEW

File: packages/dds/merge-tree/src/test/client.localReferenceFarm.spec.ts [hunk 1]
```diff
@@ -22,8 +22,8 @@ import { TestClientLogger } from "./testClientLogger";
 
 const defaultOptions: Record<"initLen" | "modLen", IConfigRange> & IMergeTreeOperationRunnerConfig =
 	{
-		initLen: { min: 2, max: 4 },
-		modLen: { min: 1, max: 8 },
+		initLen: { min: 2, max: 256 },
+		modLen: { min: 1, max: 256 },
```
Comment (connorskees) [inline packages/dds/merge-tree/src/test/client.localReferenceFarm.spec.ts:26]: See commit message in https://github.com/microsoft/FluidFramework/pull/17675/commits/d4ebf222cec53215ecf4ce781e5e11d8c87cc541

Comment (Abe27342) [reply packages/dds/merge-tree/src/test/client.localReferenceFarm.spec.ts:26]: neat!

File: packages/dds/test-dds-utils/src/ddsFuzzHarness.ts [hunk 2]
```diff
@@ -1005,20 +1013,27 @@ function runTest<TChannelFactory extends IChannelFactory, TOperation extends Bas
...
 	const itFn = options.only.has(seed) ? it.only : options.skip.has(seed) ? it.skip : it;
-	itFn(`seed ${seed}`, async () => {
+	itFn(`seed ${seed}`, async function () {
+		// 10 seconds per test should be quite a bit more than is necessary, but
+		// a timeout during minimization can cause bad UX because it obfuscates
+		// the actual error
+		//
+		// it should be noted that if a timeout occurs during minimization, the
+		// intermediate results are not lost and will still be written to the file
+		this.timeout(options.skipMinimization ? 2000 : 10_000);
```
Comment (Abe27342) [inline packages/dds/test-dds-utils/src/ddsFuzzHarness.ts:1023]: we should reuse the same check we do below. probably put it into a local

```
const shouldMinimize = !options.skipMinimization && saveInfo && !inCi
```

File: packages/runtime/test-runtime-utils/src/mocksForReconnection.ts [hunk 3]
```diff
@@ -3,7 +3,7 @@
  * Licensed under the MIT License.
  */
 
-import { v4 as uuid } from "uuid";
+import { v4 as uuidv4 } from "uuid";
```
Comment (connorskees) [inline packages/runtime/test-runtime-utils/src/mocksForReconnection.ts:6]: I will just revert all changes to this file. Ignore for review

**Review by Abe27342 — APPROVED** [review]

File: packages/dds/test-dds-utils/src/minification.ts [hunk 4]
```diff
@@ -190,12 +199,18 @@ export class FuzzTestMinimizer<
 				return false;
 			}
 
+			// reproduce based on the final line+col of the error
+			const message = error.stack
+				?.split("\n")
+				.map((s) => s.trim())
+				.find((s) => s.startsWith("at"));
```
Comment (connorskees) [inline packages/dds/test-dds-utils/src/minification.ts:206]: We will probably want to improve this check in the future. It solves the problem of e.g. string length being included in the error message, but introduces a new problem in that our crashes tend to come from the same assertion function.

Comment (Abe27342) [reply packages/dds/test-dds-utils/src/minification.ts:206]: I debated commenting about this actually. If you want to change it now, I'd make the policy injectable, e.g.

```ts
interface DDSFuzzSuiteOptions {
    minimization: {
        skip?: boolean;
        areEquivalentErrors: (original: Error, new: Error) => boolean
    }
}
```

For the `assert` concern, might be worth special-casing it and automatically counting different error codes as different.

Comment (connorskees) [reply packages/dds/test-dds-utils/src/minification.ts:206]: I am planning to merge this now as-is, but follow up with that change. I'm dogfooding these minimization changes in next to debug obliterate reconnect, and this change being in next will let me clean up a lot of my local changes.

**Pull request merged**